include/toolchain/common.h

```diff
@@ -144,4 +144,48 @@
 #define BUILD_ASSERT_MSG(EXPR, MSG) BUILD_ASSERT(EXPR)
 #endif
 
+/*
+ * This is meant to be used in conjonction with __in_section() and similar
+ * where scattered structure instances are concatened together by the linker
+ * and walked by the code at run time just like a contiguous array of such
+ * structures.
+ *
+ * Assemblers and linkers may insert alignment padding by default whose
+ * size is larger than the natural alignment for those structures when
+ * gathering various section segments together, messing up the array walk.
+ * To prevent this, we need to provide an explicit alignment not to rely
+ * on the default that might just work by luck.
+ *
+ * Alignment statements in  linker scripts are not sufficient as
+ * the assembler may add padding by itself to each segment when switching
+ * between sections within the same file even if it merges many such segments
+ * into a single section in the end.
+ */
+#define __decl_align(type) __aligned(__alignof(type)) type
+
+/*
+ * Convenience helper combining __in_section() and __decl_align().
+ * The section name is the struct type prepended with an underscore.
+ * The subsection is "static" and the subsubsection is the variable name.
+ */
+#define declare_struct_in_section(struct_type, name) \
+	__decl_align(struct struct_type) name \
+	__in_section(_##struct_type, static, name) __used
+
+/*
+ * Itterator for structure instances gathered by declare_struct_in_section().
+ * The linker must provide a _<struct_type>_list_start symbol and a
+ * _<struct_type>_list_end symbol to mark the start and the end of the
+ * list of struct objects to iterate over.
+ */
+#define iterate_struct_from_section(struct_type, iterator) \
```

Comment:
Similarly, common Zephyrish would name this "Z_STRUCT_SECTION_FOREACH" (or per my suggestion: Z_ITERABLE_STATIC_FOREACH).